backend/lambdas/api/authorizer/authorizer/handlers.py

```diff
@@ -165,25 +170,83 @@ def _verify_claims(claims: dict):
         raise Exception("Unauthorized")
 
 
-def _get_user(email: str) -> User:
-    # Create the user, if necessary
-    user, created = User.objects.get_or_create(email=email, defaults={"scope": DEFAULT_SCOPE})
-    if created:
-        # User was created so create their self group as well
-        Group.create_self_group(user)
-
-    if user.deleted:
-        # User record is soft deleted, return None so that auth fails
-        return None
+def _get_update_or_create_user(email: str) -> User:
+    """
+    Attempt to get a user based on email.
+    If user does not exist, try to match based on EMAIL_DOMAIN_ALIASES, and update email if successful.
+    If a user is still not found, create a new one.
+    """
+    user = _get_user(email)
 
-    # Set the last login timestamp
+    # If user is found, return the user (unless the user is soft-deleted)
+    if user:
+        # If user is soft-deleted, return None so that auth fails
+        if user.deleted:
+            return None
+        else:
+            return user
+
+    if EMAIL_DOMAIN_ALIASES:
+        email_domain_aliases = json.loads(EMAIL_DOMAIN_ALIASES)
+        email_local_part = email.split("@")[0]
+        email_domain = email.split("@")[1]
+
+        # Check if any aliases exist for domain, and attempt to find a match
+        for alias in email_domain_aliases:
+            if alias["new_domain"] == email_domain:
+                for old_domain in alias["old_domains"]:
+                    if alias.get("email_transformation"):
+                        transformed_email_local_part = re.sub(
+                            alias["email_transformation"]["new_email_regex"],
+                            alias["email_transformation"]["old_email_expr"],
+                            email_local_part,
+                        )
+                        old_email = f"{transformed_email_local_part}@{old_domain}"
+                    else:
+                        old_email = f"{email_local_part}@{old_domain}"
+
+                    user = _get_user(old_email)
+
+                    # If a user is found with an old email (and was not soft-deleted), update the email and return user
+                    if user and not user.deleted:
+                        user.email = email
+                        user.save()
+                        return user
+
+    # Create the user since no match has been found at this point
+    return _create_user(email)
```

Comment:
I think this might need to go inside a database transaction. The old use of `User.objects.get_or_create` was safe but I think this now isn't without a transaction.

Reply:
https://docs.djangoproject.com/en/3.2/topics/db/transactions/